File: ra-idp/src/oauth2/domain/authorization-code.ts
/**
 * Layer 3 — Application Logic
 *
 * 認可コードのドメインモデル。
 * RFC 9700 §4.10 に従う「単一使用」と TTL 制約を含む。
 */

import { randomBytes } from 'crypto'
import { AuthorizationCodeSchema, type AuthorizationCode } from '../../spec-bindings/schemas'
import { OAuthError } from '../protocol/oauth-error'

export type { AuthorizationCode }

const CODE_BYTES = 32 // 256 ビット

export function generateAuthorizationCode(input: {
  authorization_request_id: string
  client_id: string
  sub: string
  scopes: string[]
  redirect_uri: string
  code_challenge: string
  code_challenge_method: 'S256'
  nonce?: string
  auth_time: number
  ttl_seconds?: number
  now?: Date
}): AuthorizationCode {
  const ttl = input.ttl_seconds ?? 60
  const now = input.now ?? new Date()
  return AuthorizationCodeSchema.parse({
    code: randomBytes(CODE_BYTES).toString('base64url'),
    authorization_request_id: input.authorization_request_id,
    client_id: input.client_id,
    sub: input.sub,
    scopes: input.scopes,
    redirect_uri: input.redirect_uri,
    code_challenge: input.code_challenge,
    code_challenge_method: input.code_challenge_method,
    nonce: input.nonce,
    auth_time: input.auth_time,
    issued_at: now.toISOString(),
    expires_at: new Date(now.getTime() + ttl * 1000).toISOString(),
  })
}

export function isExpired(code: AuthorizationCode, now: Date = new Date()): boolean {
  return now.getTime() >= Date.parse(code.expires_at)
}

export function isRedeemed(code: AuthorizationCode): boolean {
  return !!code.redeemed_at
}

export function markRedeemed(code: AuthorizationCode, now: Date = new Date()): AuthorizationCode {
  if (isRedeemed(code)) {
    throw new OAuthError('invalid_grant', '認可コードはすでに使用されています')
  }
  return AuthorizationCodeSchema.parse({ ...code, redeemed_at: now.toISOString() })
}

File: ra-idp-go/internal/oauth2/domain/authorization_code.go
// 認可コードのドメインモデル。RFC 9700 §4.10 の単一使用 + TTL。
package domain

import (
	"crypto/rand"
	"encoding/base64"
	"time"

	"ra-idp-go/internal/shared/spec"
)

const authCodeBytes = 32

type AuthorizationCodeInput struct {
	TenantID               string
	AuthorizationRequestID string
	ClientID               string
	Sub                    string
	Scopes                 []string
	RedirectURI            string
	CodeChallenge          string
	CodeChallengeMethod    spec.CodeChallengeMethod
	Nonce                  *string
	AuthTime               int64
	TTLSeconds             int
	Now                    time.Time
}

func GenerateAuthorizationCode(in AuthorizationCodeInput) (*spec.AuthorizationCodeRecord, error) {
	if in.TenantID == "" {
		in.TenantID = spec.DefaultTenantID
	}
	if in.TTLSeconds == 0 {
		in.TTLSeconds = 60
	}
	if in.Now.IsZero() {
		in.Now = time.Now().UTC()
	}
	b := make([]byte, authCodeBytes)
	if _, err := rand.Read(b); err != nil {
		return nil, err
	}
	rec := &spec.AuthorizationCodeRecord{
		Code:                   base64.RawURLEncoding.EncodeToString(b),
		TenantID:               in.TenantID,
		AuthorizationRequestID: in.AuthorizationRequestID,
		ClientID:               in.ClientID,
		Sub:                    in.Sub,
		Scopes:                 in.Scopes,
		RedirectURI:            in.RedirectURI,
		CodeChallenge:          in.CodeChallenge,
		CodeChallengeMethod:    in.CodeChallengeMethod,
		Nonce:                  in.Nonce,
		AuthTime:               in.AuthTime,
		State:                  spec.AuthCodeRecordIssued,
		IssuedAt:               in.Now,
		ExpiresAt:              in.Now.Add(time.Duration(in.TTLSeconds) * time.Second),
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	return rec, nil
}

func IsCodeExpired(rec *spec.AuthorizationCodeRecord, now time.Time) bool {
	if now.IsZero() {
		now = time.Now()
	}
	return !now.Before(rec.ExpiresAt)
}

func IsCodeRedeemed(rec *spec.AuthorizationCodeRecord) bool {
	return rec.RedeemedAt != nil
}
